test/functional/rpc_getblockstats.py

```diff
@@ -12,130 +12,80 @@
-            height += 1
-
-        to_dump = {
-            'blocks': blocks,
-            'mocktime': int(mocktime),
-            'stats': self.expected_stats,
-        }
-        with open(filename, 'w', encoding="utf8") as f:
-            json.dump(to_dump, f, sort_keys=True, indent=2)
-
-    def load_test_data(self, filename):
-        with open(filename, 'r', encoding="utf8") as f:
-            d = json.load(f)
-            blocks = d['blocks']
-            mocktime = d['mocktime']
-            self.expected_stats = d['stats']
-
-        # Set the timestamps from the file so that the nodes can get out of Initial Block Download
-        self.nodes[0].setmocktime(mocktime)
-        self.sync_all()
+    def generate_test_data(self):
+        TRANSACTION_AMOUNTS_BTC = [10, 10, 10, 1]
 
-        for b in blocks:
-            self.nodes[0].submitblock(b)
+        wallet = MiniWallet(self.nodes[0])
 
+        self.generate(wallet, COINBASE_MATURITY + 1)
 
-    def run_test(self):
-        test_data = os.path.join(TESTSDIR, self.options.test_data)
-        if self.options.gen_test_data:
-            self.generate_test_data(test_data)
-        else:
-            self.load_test_data(test_data)
+        external_address = self.nodes[0].get_deterministic_priv_key().address
+        external_script = address_to_scriptpubkey(external_address)
+
+        for i, amount_btc in enumerate(TRANSACTION_AMOUNTS_BTC):
+            amount_satoshis = int(amount_btc * COIN)
+            wallet.send_to(
+                from_node=self.nodes[0],
+                scriptPubKey=external_script,
+                amount=amount_satoshis,
+            )
+            if i == 0:
+                self.generate(wallet, 1)
 
+        self.sync_all()
+        self.generate(wallet, 1)
+
+    def run_test(self):
+        self.generate_test_data()
         self.sync_all()
         stats = self.get_stats()
 
-        # Make sure all valid statistics are included but nothing else is
-        expected_keys = self.expected_stats[0].keys()
-        assert_equal(set(stats[0].keys()), set(expected_keys))
+        # Make sure all stats have the same set of keys
+        expected_keys = set(stats[0].keys())
+        for stat in stats[1:]:
+            assert_equal(set(stat.keys()), expected_keys)
 
         assert_equal(stats[0]['height'], self.start_height)
         assert_equal(stats[self.max_stat_pos]['height'], self.start_height + self.max_stat_pos)
 
         for i in range(self.max_stat_pos+1):
             self.log.info('Checking block %d' % (i))
-            assert_equal(stats[i], self.expected_stats[i])
 
-            # Check selecting block by hash too
-            blockhash = self.expected_stats[i]['blockhash']
+            blockhash = stats[i]['blockhash']
             stats_by_hash = self.nodes[0].getblockstats(hash_or_height=blockhash)
-            assert_equal(stats_by_hash, self.expected_stats[i])
+            assert_equal(stats_by_hash, stats[i])
```

Comment:
I think this also have a similar problem I pointed out before: https://github.com/bitcoin/bitcoin/pull/33184/files#r2567593821
The original test tests the result of getblockstats RPC over pre defined expected data. But the new test test the result of getblockstats RPC over the result of getblockstats RPC with blockhash specified. These are totally different test I think.